### PR TITLE
Pastey crate Traceability using LOBSTER

### PR DESCRIPTION
**Pull request by bharatGoswami8**
### Add Pastey Crate Traceability: Complete FMEA, Design Documentation, and Test Coverage

- Build system updated (score_tooling, Sphinx integration)
- Requirements management (system, feature, component levels)
- Architectural and static design documentation with PlantUML diagrams
- Failure Mode and Effects Analysis (FMEA) with Fault Tree Analysis (FTA)
- Test cases with LOBSTER traceability comments
- Dependable element definition with ASIL B integrity level

**Test Results**
//docs/pastey/docs:aou_test - PASSED
//docs/pastey/docs:pastey - PASSED
//docs/pastey/docs:pastey_dependability_analysis - PASSED
//docs/pastey/docs/component:pastey_component - PASSED
//docs/pastey/docs/requirement:*_test - PASSED 
//docs/pastey/tests:pastey_test -  PASSED (8 test cases)

**Documentation Generation**
Sphinx documentation builds successfully with PlantUML diagrams rendered
LOBSTER report generated with full traceability links
HTML and RST output verified

LOBSTER Report is attached here - 
[pastey_index_report.html](https://github.com/user-attachments/files/28588652/pastey_index_report.html)

### CODE REVIEW

**Comment by bharatGoswami8** [comment]
End-to-end traceability has been established using the `LOBSTER` tool within the score-crates repository. 

Although unit test results are not currently displayed in the generated reports due to a known limitation in Rust test result parsing (https://github.com/eclipse-score/tooling/issues/248 same issue created on lobster tool side as well).

Report is attached with PR description, we will enable the GitHub workflow for documentation upload/update with PR with a separate PR.

File: docs/pastey/docs/safety_analysis/root_causes/compilation_errors_fta.puml [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+ *
+ * See the NOTICE file(s) distributed with this work for additional
+ * information regarding copyright ownership.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Apache License Version 2.0 which is available at
+ * https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ ********************************************************************************/
```
Comment (arvindr19) [inline docs/pastey/docs/safety_analysis/root_causes/compilation_errors_fta.puml]: Nitpick: Other `.puml` files use `#`; this one uses `*`.

File: docs/pastey/docs/safety_analysis/root_causes/invalid_syntax_position_fta.puml [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+ *
+ * See the NOTICE file(s) distributed with this work for additional
+ * information regarding copyright ownership.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Apache License Version 2.0 which is available at
+ * https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ ********************************************************************************/
```
Comment (arvindr19) [inline docs/pastey/docs/safety_analysis/root_causes/invalid_syntax_position_fta.puml]: same as above

File: docs/pastey/docs/safety_analysis/failure_modes.trlc [hunk 3]
```diff
@@ -0,0 +1,35 @@
+/********************************************************************************
```
Comment (arvindr19) [inline docs/pastey/docs/safety_analysis/failure_modes.trlc:1]: same as above.

File: docs/pastey/docs/safety_analysis/root_causes/compilation_errors_fta.puml [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+@startuml compilation_errors_fta
+
+!include fta_metamodel.puml
+
+$TopEvent("Macro fails to synthesize identifier compilation error", "PasteyFailureModes.CompilationErrors")
+
+$OrGate("Gate1", "TopEvent")
+
+$BasicEvent("Compilation error due to empty token stream", "PasteyAoU.PasteyRustCompilerCertified", "Gate1")
+
```
Comment (arvindr19) [inline docs/pastey/docs/safety_analysis/root_causes/compilation_errors_fta.puml]: question: `aou.tlrc` has two errors. Where did this come from?

File: docs/pastey/docs/safety_analysis/root_causes/invalid_syntax_position_fta.puml [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+
+@startuml invalid_syntax_position_fta
+
+!include fta_metamodel.puml
+
+$TopEvent("paste! macro expands in syntactically invalid position", "PasteyFailureModes.InvalidTokenPlacement")
+
+$OrGate("Gate1", "TopEvent")
+
+$BasicEvent("User passes empty interpolation block to paste! macro", "PasteyAoU.PasteyUsedInValidSyntaxPositions", "Gate1")
```
Comment (arvindr19) [inline docs/pastey/docs/safety_analysis/root_causes/invalid_syntax_position_fta.puml]: same as above

File: docs/pastey/docs/safety_analysis/aou.trlc [hunk 6]
```diff
@@ -26,5 +26,5 @@ ScoreReq.AoU PasteyUsedInValidSyntaxPositions {
     safety = ScoreReq.Asil.B
     note = "The pastey crate emits raw tokens into the surrounding code without additional context checks. Using it in invalid positions will result in compilation errors or undefined behavior."
```
Comment (arvindr19) [inline docs/pastey/docs/safety_analysis/aou.trlc]: `undefined behavior` remove?

File: docs/pastey/tests/pastey_test.rs [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+    #[allow(non_snake_case)]
+    fn [<MyFunction:lower_camel>]() -> &'static str { "lowerCamelCase" }
+}
+
+#[test]
+fn test_lower_camel_modifier() {
+    // lobster-trace: PasteyComReq.REQ_COMP_PASTEY_006
+    assert_eq!(myFunction(), "lowerCamelCase");
+}
+
```
Comment (arvindr19) [inline docs/pastey/tests/pastey_test.rs:85]: Aren't we missing some requirements here? Is that intentional?
7,8,10 & 11

File: docs/pastey/docs/safety_analysis/root_causes/compilation_errors_fta.puml (hunk 4 above)
Comment (bharatGoswami8) [reply docs/pastey/docs/safety_analysis/root_causes/compilation_errors_fta.puml]: Assumption of use are not error it is assumptions about how user going to use the component.

File: docs/pastey/docs/safety_analysis/root_causes/invalid_syntax_position_fta.puml (hunk 5 above)
Comment (bharatGoswami8) [reply docs/pastey/docs/safety_analysis/root_causes/invalid_syntax_position_fta.puml]: Updated above.

File: docs/pastey/docs/safety_analysis/aou.trlc (hunk 6 above)
Comment (bharatGoswami8) [reply docs/pastey/docs/safety_analysis/aou.trlc]: yes and it should be removed from note also as `pastey` compile time macro crate.

File: docs/pastey/docs/safety_analysis/failure_modes.trlc (hunk 3 above)
Comment (bharatGoswami8) [reply docs/pastey/docs/safety_analysis/failure_modes.trlc:1]: All the `trlc` has LICENSE header in this format.

File: docs/pastey/tests/pastey_test.rs (hunk 7 above)
Comment (bharatGoswami8) [reply docs/pastey/tests/pastey_test.rs:85]: No, it was not intentional, while cleanup few test cases were removed added back,

File: docs/pastey/docs/safety_analysis/root_causes/compilation_errors_fta.puml (hunk 1 above)
Comment (bharatGoswami8) [reply docs/pastey/docs/safety_analysis/root_causes/compilation_errors_fta.puml]: starting with `#` is only for BUILD file and `/*` for other files, for puml it should be `'` comment with this.
I am not sure how for this it was working with `/*` but updated for all the puml now.

File: docs/pastey/docs/safety_analysis/root_causes/invalid_syntax_position_fta.puml (hunk 2 above)
Comment (bharatGoswami8) [reply docs/pastey/docs/safety_analysis/root_causes/invalid_syntax_position_fta.puml]: starting with `#` is only for BUILD file and `/*` for other files, for puml it should be `'` comment with this.
I am not sure how for this it was working with `/*` but updated for all the puml now.

File: MODULE.bazel.lock
Comment (rpreddyhv) [inline MODULE.bazel.lock:1]: question: Does this need to be in repo? Is it something similar to `Cargo.lock` in `rust` world?

File: third_party/sphinx/BUILD
Comment (rpreddyhv) [inline third_party/sphinx/BUILD:1]: suggestion: Looks like this has nothing to do with `pastey` crate traceability. You could consider moving all toolchain updates to separate PR if possible.

File: docs/pastey/tests/pastey_test.rs
Comment (rpreddyhv) [inline docs/pastey/tests/pastey_test.rs:1]: suggestion: I know the tests & their types are already organised well. But, if it possible to group test related types/setup together.

Comment (rpreddyhv) [inline docs/pastey/tests/pastey_test.rs:1]: question: I see the tests exist in the https://github.com/AS1100K/pastey/tree/master/paste-compat/tests for pastey crate. How these tests are different from those in the upstream repo? Are these tests here just for traceability?

File: docs/pastey/docs/BUILD [hunk 8]
```diff
@@ -11,12 +11,34 @@
...
+    arch_design = "//docs/pastey/docs/design:pastey_design",
+    fmea = ["//docs/pastey/docs/safety_analysis:pastey_fmea"],
+)
+
+dependable_element(
+    name = "pastey",
+    integrity_level = "B",
+    assumptions_of_use = [":aou"],
+    requirements = [
+        "//docs/pastey/docs/requirement:feature_requirements",
```
Comment (rpreddyhv) [inline docs/pastey/docs/BUILD:38]: question: If `component:pastey_component` already including `component_requirements` & `feature_requirements`, do we need the `feature_requirements` again here?

File: docs/pastey/docs/safety_analysis/root_causes/compilation_errors_fta.puml [hunk 9]
```diff
@@ -0,0 +1,21 @@
...
+' SPDX-License-Identifier: Apache-2.0
+@startuml compilation_errors_fta
+
+!include fta_metamodel.puml
+
+$TopEvent("Macro fails to synthesize identifier compilation error", "PasteyFailureModes.CompilationErrors")
+
+$OrGate("Gate1", "TopEvent")
+
+$BasicEvent("Compilation error due to empty token stream", "PasteyAoU.PasteyRustCompilerCertified", "Gate1")
```
Comment (rpreddyhv) [inline docs/pastey/docs/safety_analysis/root_causes/compilation_errors_fta.puml]: question: How ""Compilation error due to empty token stream" is related to ""PasteyAoU.PasteyRustCompilerCertified"?

File: docs/pastey/docs/safety_analysis/aou.trlc
Comment (rpreddyhv) [inline docs/pastey/docs/safety_analysis/aou.trlc:21]: question: There is no mitigation for `PasteyRustCompilerCertified`? Is that justified elsewhere?

**Review by rpreddyhv — COMMENTED** [review]
👏 I appreciate the efforts put into the traceability of first SCORE crate. This becomes a reference for traceability of other such crates, though it is macro only crate.

I have few questions & suggestions. No critical findings from my side.

File: docs/pastey/docs/BUILD (hunk 8 above)
Comment (bharatGoswami8) [reply docs/pastey/docs/BUILD:38]: yes, we do need to mention here because this requires checking traceability of feature required.

File: docs/pastey/tests/pastey_test.rs
Comment (bharatGoswami8) [reply docs/pastey/tests/pastey_test.rs:1]: yes, it is just for traceability.

Comment (bharatGoswami8) [reply docs/pastey/tests/pastey_test.rs:1]: I think tests and types are organised well already.

File: third_party/sphinx/BUILD
Comment (bharatGoswami8) [reply third_party/sphinx/BUILD:1]: As we already have separate commits for each infra-side change, and all those changes are related to enabling traceability report generation for `pastey`, I think keeping them in the same PR makes sense. This also avoids the need to justify dependency requirements across multiple separate PRs.

File: MODULE.bazel.lock
Comment (bharatGoswami8) [reply MODULE.bazel.lock:1]: it is compiler generated so yes, it is like `cargo.lock `file and this file update required because repo has bazel workflow check.

File: docs/pastey/docs/safety_analysis/root_causes/compilation_errors_fta.puml (hunk 9 above)
Comment (bharatGoswami8) [reply docs/pastey/docs/safety_analysis/root_causes/compilation_errors_fta.puml]: This is general assumption of use for user, so we have mentioned in this first root cause analysis, if we don't link then we will get failure in bazel test for traceability.

File: docs/pastey/docs/safety_analysis/aou.trlc
Comment (bharatGoswami8) [reply docs/pastey/docs/safety_analysis/aou.trlc:21]: This is general assumption of use for user so no `mitigates`

File: docs/pastey/docs/design/architectural_design.rst
Comment (aschemmel-tech) [inline docs/pastey/docs/design/architectural_design.rst:1]: If you want to document design: please follow the folder structure as defined in https://eclipse-score.github.io/score/main/contribute/general/folder.html for components - folder name should be "architecture" or "detailed_design". Looking at the provided puml, I would say it's "architecture"

File: docs/pastey/docs/design/architectural_design.rst [hunk 10]
```diff
@@ -0,0 +1,23 @@
+Pastey Architectural Design
```
Comment (aschemmel-tech) [inline docs/pastey/docs/design/architectural_design.rst]: I am not sure we need this, but if we do I recommend using the templates for architecture in https://github.com/eclipse-score/module_template/blob/main/score/component_example/docs/architecture/index.rst

File: docs/pastey/docs/safety_analysis/failure_modes.trlc [hunk 11]
```diff
@@ -0,0 +1,35 @@
...
+ * SPDX-License-Identifier: Apache-2.0
+ ********************************************************************************/
+package PasteyFailureModes
+
+import ScoreReq
+
+ScoreReq.FailureMode CompilationErrors {
+    guideword = ScoreReq.GuideWord.LossOfFunction
+    description = "The paste! macro shall fail to synthesize an identifier when the token stream inside the interpolation block is empty or contains only whitespace."
+    failureeffect = "Compilation fails with a macro expansion error, preventing the build from succeeding."
```
Comment (aschemmel-tech) [inline docs/pastey/docs/safety_analysis/failure_modes.trlc]: failure effect is not safety relevant as the code will never run on target, so failure mode (and also the respective AoU) can be removed

File: docs/pastey/docs/safety_analysis/failure_modes.trlc [hunk 12]
```diff
@@ -0,0 +1,35 @@
...
+    safety = ScoreReq.Asil.B
+    version = 1
+    potentialcause = "Caller passes an empty or whitespace-only interpolation block to the paste! macro."
+    interface = "pastey_pub.paste_macro"
+}
+
+ScoreReq.FailureMode InvalidTokenPlacement {
+    guideword = ScoreReq.GuideWord.UnintendedFunction
+    description = "The paste! macro shall expand tokens in a syntactically invalid position causing the surrounding Rust code to fail to compile."
+    failureeffect = "Build failure due to invalid token placement in generated code."
```
Comment (aschemmel-tech) [inline docs/pastey/docs/safety_analysis/failure_modes.trlc]: failure effect is not safety relevant as the code will never run on target, so failure mode (and also the respective AoU) can be removed

File: docs/pastey/docs/safety_analysis/aou.trlc
Comment (aschemmel-tech) [inline docs/pastey/docs/safety_analysis/aou.trlc:18]: This AoU misses a failure mode which causes it.

File: docs/pastey/tests/pastey_test.rs
Comment (aschemmel-tech) [inline docs/pastey/tests/pastey_test.rs:1]: could not find a test covering FEAT_PASTEY_001 - how could a integration test be done?

File: docs/pastey/tests/pastey_test.rs [hunk 13]
```diff
@@ -0,0 +1,137 @@
...
+    assert_eq!(MyHttpServer(), "MyHttpServer");
+}
+
+paste! {
+    fn [<replace_test:replace(test, demo)>]() -> &'static str { "replace_demo" }
+}
+
+#[test]
+fn test_replace_modifier() {
+    // lobster-trace: PasteyComReq.REQ_COMP_PASTEY_008
```
Comment (aschemmel-tech) [inline docs/pastey/tests/pastey_test.rs:103]: I cannot see this here, but do you also test the negative outcomes of this requirement? This is a general question also applicable maybe to other requirements.

File: docs/pastey/docs/safety_analysis/failure_modes.trlc [hunk 14]
```diff
@@ -0,0 +1,35 @@
...
+
+import ScoreReq
+
+ScoreReq.FailureMode CompilationErrors {
+    guideword = ScoreReq.GuideWord.LossOfFunction
+    description = "The paste! macro shall fail to synthesize an identifier when the token stream inside the interpolation block is empty or contains only whitespace."
+    failureeffect = "Compilation fails with a macro expansion error, preventing the build from succeeding."
+    safety = ScoreReq.Asil.B
+    version = 1
+    potentialcause = "Caller passes an empty or whitespace-only interpolation block to the paste! macro."
```
Comment (aschemmel-tech) [inline docs/pastey/docs/safety_analysis/failure_modes.trlc]: now that I think of testing coverage, this would also be a valid test case.

**Review by aschemmel-tech — CHANGES_REQUESTED** [review]
See inline comments

File: docs/pastey/docs/design/architectural_design.rst (hunk 10 above)
Comment (bharatGoswami8) [reply docs/pastey/docs/design/architectural_design.rst]: We need this for LOBSTER tool traceability report generation.

File: docs/pastey/tests/pastey_test.rs
Comment (bharatGoswami8) [reply docs/pastey/tests/pastey_test.rs:1]: Feature requirement is link with component requirement and as we discussed feature requirement for Rust crate is just general brief about crate as feature and we are linking that same requirement with component, so traceability is covering using component requirement.

If we need to mention explicitly then I can add with each test case.

File: docs/pastey/docs/safety_analysis/failure_modes.trlc (hunk 14 above)
Comment (bharatGoswami8) [reply docs/pastey/docs/safety_analysis/failure_modes.trlc]: it is already there on upstream - https://github.com/AS1100K/pastey/blob/master/src/lib.rs#L706C1-L707C1

`  fn test_error_empty_macro_result() {}`

File: docs/pastey/docs/design/architectural_design.rst
Comment (bharatGoswami8) [reply docs/pastey/docs/design/architectural_design.rst:1]: Updated as per shared reference.

File: docs/pastey/docs/design/architectural_design.rst (hunk 10 above)
Comment (bharatGoswami8) [reply docs/pastey/docs/design/architectural_design.rst]: Updated the detail.

File: docs/pastey/docs/safety_analysis/failure_modes.trlc (hunk 11 above)
Comment (bharatGoswami8) [reply docs/pastey/docs/safety_analysis/failure_modes.trlc]: Removed this and AoU for same.

File: docs/pastey/docs/safety_analysis/failure_modes.trlc (hunk 12 above)
Comment (bharatGoswami8) [reply docs/pastey/docs/safety_analysis/failure_modes.trlc]: Removed this and AoU for same.

File: docs/pastey/docs/safety_analysis/aou.trlc
Comment (bharatGoswami8) [reply docs/pastey/docs/safety_analysis/aou.trlc:18]: Updated.

File: docs/pastey/tests/pastey_test.rs (hunk 13 above)
Comment (bharatGoswami8) [reply docs/pastey/tests/pastey_test.rs:103]: yes at upstream we have all the test cases and coverage , this test cases just for component requirement trackability.
https://github.com/AS1100K/pastey/blob/master/

Comment (aschemmel-tech) [reply docs/pastey/tests/pastey_test.rs:103]: strongly recommend to also add the (as you say existing) tests covering the negative outcome to this requirement (and the others where applicable).

File: docs/pastey/docs/safety_analysis/failure_modes.trlc (hunk 14 above)
Comment (aschemmel-tech) [reply docs/pastey/docs/safety_analysis/failure_modes.trlc]: Then I would recommend to add this as a requirement.

File: docs/pastey/tests/pastey_test.rs
Comment (aschemmel-tech) [reply docs/pastey/tests/pastey_test.rs:1]: let's leave this open for the moment

File: docs/pastey/tests/pastey_test.rs (hunk 13 above)
Comment (bharatGoswami8) [reply docs/pastey/tests/pastey_test.rs:103]: Added the `doc test `in rust which can check compiler failure / error for negative case and error requirement.

File: docs/pastey/docs/safety_analysis/failure_modes.trlc (hunk 14 above)
Comment (bharatGoswami8) [reply docs/pastey/docs/safety_analysis/failure_modes.trlc]: `REQ_COMP_PASTEY_013 ` added.

**Review by aschemmel-tech — APPROVED** [review]
my findings resolved
